### PR TITLE
Changes to auto mode in EarlyStopping callback

### PR DESCRIPTION
**Pull request by SuryanarayanaY**
In current implementation of EarlyStopping callback, If the argument mode is not mentioned in either min or max it fall backs to auto mode. In auto mode if the monitored metric name not ends with 'acc' or 'accuracy' ot 'auc' then the mode will fall back to mode= 'min' which is problematic. For example if the monitored metric is 'TruePositive' or 'FalsePositive' for both cases it will monitor whether the metric is decreasing or not by minimum change configured in callback.This makes callback not useful at all for this cases.

Similarly there are metrics in [TF-ranking](https://www.tensorflow.org/ranking/api_docs/python/tfr/keras/metrics), which by default fall to min mode if user left the mode to auto.

Relevant issues from [TF-Forum](https://discuss.tensorflow.org/t/confusion-in-earlystopping-mode-argument/14080) and [SO](https://stackoverflow.com/questions/69744405/tensorflow-earlystopping-stops-too-early).

IMO, its better to leave the metrics `mode` to be defined by user. Otherwise we need to list all the metrics that should have higher value(like `accuracy`) and lower values(like `loss`) to efficiently use this auto mode. This may be problematic for custom metrics.

Maybe we can leave a warning/exception to the user if metrics other than 'acc' or 'accuracy' ot 'auc' or used as monitored metric something like below.

### CODE REVIEW

**Comment by SuryanarayanaY** [comment]
The tests are failing for the case of `metric = unknown` and `mode = unknown`.

https://github.com/keras-team/keras/blob/1f0a013b02bbafc8a016353a0e60ba94895fea07/keras/callbacks/early_stopping_test.py#L34

We remove this case because with the proposed changes of this PR this case will fail.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR. I suggest we change the docstring instead, and say,

```
in `"auto"` mode, the direction is automatically inferred from the name
of the monitored quantity: it will default to 'min' mode unless the monitored quantity is on of {"accuracy", "auc"}, in which case it uses 'max' mode.
```

There may be a way to make the inference more robust but it's not simple.

**Comment by SuryanarayanaY** [comment]
@fchollet ,

 Suppose the metric is `f1-score` and user not mentioned `mode` and by default it fall backs to `"auto"` mode in which case `f1-score` metric will default to `"min"` mode. But we know `f1-score` should be higher for a better model but because of the `"auto"` mode the model will continue training since `f1-score` improves with each epoch and `EarlyStopping` callback will never called. Even if it is called for decreasing value of  `f1-score` that is still pretty bad as it will stop training for pretty bad results.

IMO, in this particular case the purpose of `EarlyStopping` not being served.Hence proposed for exception as implementation of `auto` mode to work properly is not simple.

If it is still not a good idea to raise `exception` we may amend the `warning` here if OK.

https://github.com/keras-team/keras/blob/4501d8b9bc1e6f4f10ac09fe89f61494330b98a2/keras/callbacks/early_stopping.py#L93-L95

like:

```
if mode not in ["auto", "min", "max"]:
            warnings.warn(
                f"EarlyStopping mode {mode} is unknown, fallback to auto mode.",
                "in `"auto"` mode, the direction is automatically inferred from "
                " the name of the monitored quantity:" 
                " it will default to 'min' mode unless the monitored quantity is"
                "one of {"accuracy", "auc"}, in which case it uses 'max' mode." ,
                stacklevel=2,
            )
            mode = "auto"
```

or some warning like below:

```
if mode not in ["auto", "min", "max"]:
            warnings.warn(
                f"EarlyStopping mode {mode} is unknown, fallback to auto mode.",
                "`"auto"` mode may not be suitable for your metric, hence decide"
                " the mode whether `"min"` or  `"max"`  " ,
                stacklevel=2,
            )
            mode = "auto"
```

**Comment by fchollet** [comment]
What's the full list of built-in metrics that should be in max mode?

**Comment by fchollet** [comment]
Ok, I pushed an extensive fix. Check it out.
